haystack/preview/components/routers/document_joiner.py

```diff
@@ -0,0 +1,142 @@
+    p.connect("bm25_retriever", "joiner")
+    p.connect("embedding_retriever", "joiner")
+    ```
+    """
+
+    def __init__(
+        self,
+        join_mode: str = "concatenate",
+        weights: Optional[List[float]] = None,
+        top_k: Optional[int] = None,
+        sort_by_score: bool = True,
+    ):
+        """
+        Initialize the DocumentJoiner.
+
+        :param join_mode: `concatenate` to combine documents from multiple retrievers, `merge` to aggregate scores of
+                          individual documents, `reciprocal_rank_fusion` to apply rank based scoring.
+        :param weights: A node-wise list(length of list must be equal to the number of input nodes) of weights for
+                        adjusting document scores when using the `merge` join_mode. By default, equal weight is given
+                        to each retriever score. This param is not compatible with the `concatenate` join_mode.
+        :param top_k: How many documents should be returned in the output at maximum. By default, all are returned.
+        :param sort_by_score: Whether the output list of documents should be sorted by document scores in descending order. By default the output is sorted. Documents without score are handled as if their score was -infinity.
+        """
+        if join_mode not in ["concatenate", "merge", "reciprocal_rank_fusion"]:
+            raise ValueError(f"DocumentJoiner component does not support '{join_mode}' join_mode.")
+        self.join_mode = join_mode
+        self.weights = [float(i) / sum(weights) for i in weights] if weights else None
+        self.top_k = top_k
+        self.sort_by_score = sort_by_score
+
+    @component.output_types(documents=List[Document])
+    def run(self, documents: Variadic[List[Document]]):
+        """
+        Run the DocumentJoiner. This method joins the input lists of documents into one output list based on the join_mode specified during initialization.
+
+        :param documents: An arbitrary number of lists of documents to join.
+        """
+        output_documents = []
+        if self.join_mode == "concatenate":
+            output_documents = self._concatenate(documents)
+        elif self.join_mode == "merge":
+            output_documents = self._merge(documents)
+        elif self.join_mode == "reciprocal_rank_fusion":
+            output_documents = self._reciprocal_rank_fusion(documents)
+
+        if self.sort_by_score:
+            output_documents = sorted(
+                output_documents, key=lambda doc: doc.score if doc.score is not None else -inf, reverse=True
+            )
+            if any(doc.score is None for doc in output_documents):
+                logger.info(
+                    "The `DocumentJoiner` component has received some documents with `score=None` - and was requested "
+                    "to sort the documents by score, so the `score=None` documents were sorted as if their "
+                    "score was `-infinity`."
```

Comment:
Some of the Documents DocumentJoiner got have `score=None`. It was configured to sort Documents by score, so those with `score=None` were sorted as if they has a score of `-infinity`.